Smarter default behavior for -printtoconsole

eklitzke opened the pull request:
This makes console printing smarter when `-printtoconsole` is not explicitly set. The new behavior:
 * The series of newline separators that we send to the log when starting up are only sent to `debug.log`, never to stdout.
 * Logging to stdout is set up by default when the process is not a daemon.

The motivation for this change is that I almost always run bitcoind in foreground mode when I'm debugging or doing development work, and the new behavior makes checking log statements a lot easier in that situation. I think these semantics are also more similar to the default behavior of many other Unix programs.

Sjors commented:
Concept ACK. Tested on macOS.

Needs update for `-printtoconsole` help text.

donaloconnor commented on src/util.cpp:
```diff
@@ -336,6 +336,11 @@ static std::string LogTimestampStr(const std::string &str, std::atomic_bool *fSt
 
+bool IsStdoutTty()
+{
+    return isatty(STDOUT_FILENO);
```
I know you said that MingW includes a compatible unistd.h version for Windows but on Visual Studio/msvc isatty is deprecated and replaced by _isatty. Also STDOUT_FILENO (1) is not defined. There's currently some PRs for msvc support so just a heads up that this will likely break things there. #11526 and #11873

practicalswift commented:
Strong concept ACK. Very nice usability improvement! I'm also running `bitcoind` in the foreground mode for the most part.

What about reducing the number of `\n`:s from 20 to say 5? That should be more than enough to distinguish between to subsequent executions.

donaloconnor commented:
Just a few concerns (Mainly on Windows)

1. There's now no way of turning the stdout logging off when launched in foreground. It's either on or (on by default) 
2. On Windows output will always be shown now since deamonising isn't possible?. (AFAIK it's not possible to run bitcoind as a windows service?
3. stdout isn't buffered very well (if at all?) so it can potentially have a perf penalty (especially if custom high freq custom logging is enabled)

BTW fPrintToDebugLog is set to true and not changed ever so the code can be simplified.

promag commented:
Concept ACK, cool :+1:.

eklitzke commented:
Now that I've thought about this more I don't think the `isatty()` check is necessary. The situation where it wouldn't be a TTY would be if you were running commands like:

```bash
# stdout is a pipe, not a TTY.
bitcoind | less

# stdout is a file, not a TTY.
bitcoind > redundantcopyofdebug.log
```

I don't think either of these are things people are going to actually run, but if they did they would probably expect to see output printed. The usual use of `isatty()` is actually to test for features like using console escape sequences (e.g. the ASCII "bell" character) or to tell if it makes sense to print ANSI color codes.

eklitzke commented:
New version (with squashed commit) remove the `isatty()` check and reduces the number of newlines we print from 20 to 5. I've also updated the PR description to remove the commentary about `isatty()`.

donaloconnor commented:
My understanding is now logging will not be sent to the *log file* unless -daemon is passed via command line. If that's the case won't this be unexpected and annoying behavior?

LogPrintStr only prints to one or the other.

I'm on the phone so it's difficult to check

luke-jr commented:
Not sure this makes sense. I don't know of any other software that guesses stdout vs log based on whether stdout is a tty. It seems likely to break users who do `bitcoin-qt &`

Why not just put `printtoconsole=1` in bitcoin.conf if you prefer it by default?

eklitzke commented:
@lukejr The code no longer uses `isatty()`, I removed that code. Adding `printtoconsole=1` to `bitcoin.conf` has the side effect that it completely suppresses logging to `debug.log`, which is annoying if sometimes you run bitcoin as a daemon and sometimes you run it in foreground mode.

We should probably change the `LogPrintf()` logic to log to both if both flags are set. But even in that case if you add `printtoconsole=1` to your bitcoin.conf file you'll get behavior where bitcoind would log all messages twice, with the stdout one going to `/dev/null`. This is harmless but inefficient.

eklitzke commented:
@donaloconnor Agree and I just noticed that myself. I will update this to log to both in that case, which I think is better behavior.

eklitzke commented:
@donaloconnor I also just noticed that which I think is bad behavior. Updated to make these flags independent (printing to console will not affect logging to debug.log).

practicalswift commented:
utACK 23bc0aaac8d8ac7f2b41527567d6a2329429d172

laanwj commented on src/util.cpp:
```diff
@@ -349,7 +349,7 @@ int LogPrintStr(const std::string &str)
         fflush(stdout);
     }
-    else if (fPrintToDebugLog)
+    if (fPrintToDebugLog)
```
I'm fine with making these non-exclusive, however we still need a way to disable logging to `debug.log`. Some logging setups (e.g. when using systemd) process the `-printtoconsole` output, and don't need a duplicate logging to disk.
Maybe `-nodebuglogfile`?

Sjors replied:
Maybe `-debuglogfile=0` as a special case?

practicalswift replied:
What about `-debuglogfile=/dev/null` as a special case .-)

If we are to add a separate option, then perhaps something starting with `-disable` to make it consistent with `-disablewallet`? We don't have any `bitcoind` command-line options starting with `-no` AFAIK.

Also `-nodebuglogfile` can be parsed as "node bug logfile" :-)

MeshCollider replied:
Actually -noX is interpreted by the args manager as -X=0 implicitly anyway, converted by the InterpretNegativeSetting() function. So @laanwj and @Sjors suggestions are equivalent :)

Sjors commented:
Tested 23bc0aa on macOS, still works. 

Removing the `isatty()` check makes sense; another example would be using `bitcoind | grep the_thing_I_care_about`.

@luke-jr  `bitcoin-qt` does not output anything to the console (not sure why).

Sjors commented on src/init.cpp:
```diff
@@ -814,14 +814,35 @@ static std::string ResolveErrMsg(const char * const optname, const std::string&
+    // Add newlines to the logfile to distinguish this execution from the last
+    // one; called before console logging is set up, so this is only sent to
+    // debug.log.
+    LogPrintf("\n\n\n\n\n");
```
`std::string(NUMBER_OF_BLANK_LINES_BETWEEN_SESSIONS_IN_LOG_TO_PREEMPT_BIKESHEDDING, '\n')`? :-P

Sjors commented on src/util.cpp:
```diff
@@ -349,7 +349,7 @@ int LogPrintStr(const std::string &str)
         fflush(stdout);
     }
-    else if (fPrintToDebugLog)
+    if (fPrintToDebugLog)
```
Nit: move brackets to same line?

practicalswift commented:
utACK modulo @Sjors' nits :-)

donaloconnor commented:
utACK  https://github.com/bitcoin/bitcoin/commit/23bc0aaac8d8ac7f2b41527567d6a2329429d172 though I do agree with @laanwj that disabling logging would make sense now.

eklitzke commented:
I could not resist the temptation to add `-nodebuglog` because I found the "node bug log" thing to be so funny. I added it as a "secret" option (meaning it's suppressed by default from `-help` output). You can also use `-debuglogfile=/dev/null` and `-debuglogfile=0` to disable logging. There are also Python test cases for all three methods. So I believe everyone got what they asked for :-D

Interestingly, before this change `-debuglogfile=/dev/null` did **not** work, because `FileTruncate()` does not check that its argument is a regular file. This is yet-another Unix compatibility thing that I would like Bitcoin to support better! I believe this means that master will fail if you try to log to e.g. a named pipe created with `mkfifo(1)`, which is not common but still an unfortunate bug.

One question I have about this change: I was originally going to update the `-debuglogfile` help string but realized it would require it to be re-translated. Since developers are probably the only ones who want this option it might just make sense to have a one line changelog note about the change? I'm not sure what the protocol is for this.

eklitzke commented:
I tested and you actually can use a named pipe, because the operation that's failing is a file size check in boost. But you can't use other special files, e.g. `/dev/zero` which swallows input the same was as `/dev/null`.

```bash
# this is allowed
$ echo foo > /dev/zero

# yep, it worked
$ echo $?
0

# this fails
$ ./src/bitcoind -debuglogfile=/dev/zero


************************
EXCEPTION: N5boost10filesystem16filesystem_errorE       
boost::filesystem::file_size: Operation not permitted: "/dev/zero"       
bitcoin in AppInit()       
```

Something to fix another day I guess.

laanwj commented:
> use -debuglogfile=/dev/null 

Right, what bothered me about /dev/null is that it's OS-specific. I don't think providing arbitrary device files as debug log files should be really a goal. It mentions `file` in the name :) Better to have a dedicated option for it, thanks.

eklitzke commented:
I spent way longer on this than I'd like to admit, due to strange behavior in the Bitcoin option parser.

There's a little-known feature in the option parser where `-nofoo` is the same as `-foo=0` when updating the value map. So you can't actually create an option with a name like `-nodebuglog`, because looking up the value from the map still uses the `nodebuglog` value, but the value is actually set in the map as `debuglog`. I thought this was really neat, because my last update to this diff had special cased `-nodebuglog` and `-debuglogfile=0`, so I thought I could just use `gArgs.GetBoolArg("-debuglogfile")`. That would be really nice because then I could delete a bunch of lines from my last diff and not treat those as special cases.

Nope, doesn't work. I digged in and here's where it gets really weird. The option parser uses `atoi()` to interpret boolean arguments. The thing is, the return value of `atoi()` is **undefined** if you pass it non-integer values, as there's no way to tell from the return value whether an error occurred.

Here's where it gets kind of weird. We had a test in `ResendWalletTransactionsTest()` that was setting a flag like this:

```python
class ResendWalletTransactionsTest(BitcoinTestFramework):
    def set_test_params(self):
        self.num_nodes = 1
        self.extra_args = [['--walletbroadcast=false']]
```

So what happens when the option parser sees this... is it calls `atoi("false")`. That returns 0 like we want, but only because `atoi()` happens to return false for invalid input... a fact that I want to again emphasize is undefined. On another libc implementation `atoi()` is allowed to e.g. return an uninitialized value in this situation. This also means that `-walletbroadcast=true` would **disable** wallet broadcast, for the same reason.

Since there's a test case in Bitcoin that was using a "false" flag, I think it's possible that there are other people in the wild relying on this behavior. So now the option parser allows only `0` and `false` and no other values. If people had configs that were setting values to false, this will still work. I really wanted to add C++ tests for this code, but the methods I wanted to test are defined static in the util.cpp file.

I also made it so if you use `-debuglogfile=/dev/null` on a Unix system, it won't crash. But I'm not special casing this, so it will cause bitcoind to actually open `/dev/null` and write data to it.

eklitzke commented:
Actually it looks like I can add C++ test for the option parser, I'm going to put up another commit adding that.

luke-jr commented:
`atoi` is well-defined for all null-terminated input strings.

> The strtol, strtoll, strtoul, and strtoull functions return the converted
value, if any. If no conversion could be performed, zero is returned.

eklitzke commented:
Interesting, that's not documented in the man pages on my system, but cppreference.com agrees that it must return 0 in such a case.

donaloconnor commented:
utACK - looks good

MarcoFalke commented on src/util.cpp:
```diff
@@ -423,16 +432,14 @@ void ReleaseDirectoryLocks()
-        return true;
-    return (atoi(strValue) != 0);
+    // These values mean false, anything else (including empty string) is true.
+    return strValue != "0" && strValue != "false";
```
I think this has nothing to do with `printtoconsole`. Mind to split it up into a separate pull?

eklitzke replied:
It has to do with the behavior I wanted for disabling the debug log, but you're right that it's technically orthogonal. I will split it out.

eklitzke commented:
I've split the option parsing code out into another change as Marko asked. Once that is merged I can rebase and squash this branch.

jimpo commented on test/functional/wallet_resendwallettransactions.py:
```diff
@@ -10,7 +10,7 @@
     def set_test_params(self):
         self.num_nodes = 1
-        self.extra_args = [['--walletbroadcast=false']]
+        self.extra_args = [['-walletbroadcast=0']]
```
Feels like this should also be in #12713.

Sjors replied:
This has been moved to the other commit.

jimpo commented on src/util.cpp:
```diff
@@ -193,6 +193,17 @@ fs::path GetDebugLogPath()
     return AbsPathForConfigVal(logfile);
 }
 
+bool ShouldOpenDebugLog()
```
Feels non-obvious based on the name that this would toggle the state of a global variable. Now that the bool option parsing is fixed, I don't think there's a need for this function. But even if you want to keep it, I think it would be clearer if the code in `AppInitMain` set `fPrintToDebugLog = ShouldOpenDebugLog()`.

jimpo commented on src/init.cpp:
```diff
@@ -814,14 +814,35 @@ static std::string ResolveErrMsg(const char * const optname, const std::string&
+ *
+ * @return true if we should log to stdout, false otherwise.
+ */
+static bool ShouldPrintToConsole()
```
nit: `PrintToConsoleDefault()` is seems more accurate. Otherwise the function name makes it sound like it takes `-printtoconsole` into account.

MarcoFalke replied:
Since it is only called once, you could also inline it and call the variable `const bool default_printtoconsole = !gArgs.GetBoolArg("-daemon", false);`

eklitzke replied:
It's already inlined by the compiler, there's no difference in the executable produced by GCC either way. The only reason it's still a method is because it originally called `isatty()`. If I can get #12713 merged I am happy to remove this method.

eklitzke commented:
@MarcoFalke This branch now contains two commits: the one from #12713 , and then a second commit with changes actually related to console printing. Both commits should pass the test suite.

eklitzke commented:
Updated again with changes from #12713.

Sjors reviewed (approved):
tACK c389e87 on macOS.

Would be nice if you update the RPC help though:
* `-printtoconsole` says `instead of debug.log file` which is not true

jnewbery commented on test/functional/feature_logging.py:
```diff
@@ -54,6 +58,17 @@ def run_test(self):
+
+        # just sanity check no crash here
+        self.stop_node(0)
+        self.start_node(0, ["-debuglogfile=/dev/null"])
```
Will this work on non-linux platforms? Is `os.devnull` better? https://docs.python.org/3/library/os.html#os.devnull

jnewbery reviewed (commented):
Change seems good to me. However, this causes debug log to be spewed to console when running individual functional tests (ie not through the test_runner). I think you'll need to update:

https://github.com/bitcoin/bitcoin/blob/4ad3b3c72c73d61e0a0cab541dca20acf651320d/test/functional/test_framework/test_node.py#L67

to include `-noprinttoconsole`

Also one nit in the new test.

eklitzke committed:
Add additional tests for GetBoolArg()

This is meant to be an intermediate commit to prove that the next does not
introduce any changes in the semantics of boolean option parsing.

eklitzke committed:
Track negated arguments in the argument paser.

This commit adds tracking for negated arguments. This change will be used in a
future commit that allows disabling the debug.log file using -nodebuglogfile.

ryanofsky commented on test/functional/feature_logging.py:
```diff
@@ -67,7 +67,7 @@ def run_test(self):
         # just sanity check no crash here
         self.stop_node(0)
-        self.start_node(0, ["-debuglogfile=/dev/null"])
+        self.start_node(0, ["-debuglogfile=%s" % os.devnull])
```
In commit "stop printing to console in test runner"

This change looks like it should go in the previous commit ("Enable -printtoconsole by default if bitcoind is not a daemon").

ryanofsky reviewed (commented):
utACK for the last two commits that aren't part of #12713:

- [ ] 6b95d6bd6e81030605d6b4ad42dd83cd3bc69bfd Improve boolean option parsing (1/3)
- [X] c389e870beaf5e8b84ca09ef8e7ba3c981ca6988 Enable -printtoconsole by default if bitcoind is not a daemon (2/3)
- [X] 60086ad871cb1a756c152a3d3e19fcea52660123 stop printing to console in test runner (3/3)

jnewbery commented:
utACK 60086ad871cb1a756c152a3d3e19fcea52660123. I think that commit can be squashed with the previous.

eklitzke committed:
Print to console by default when not run with -daemon

Printing to the debug log file can be disabled with -nodebulogfile

MarcoFalke commented:
Needs rebase to get rid of the already-merged diff.

MarcoFalke commented:
Marked "up for grabs"

ryanofsky reviewed (commented):
utACK ef6fa1c38e1bd115d1cce155907023d79da379d8. No changes since the last review other than squash & rebase. Note: The first two commits shown in this PR are already merged into master since 4490871ed7eb8de0b644ba8545789f2cc702434e (#12713), so they can be ignored.
